sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,362 @@
+# Admission control changes
+
+Admission control is the key security and cluster policy enforcement mechanism
+in Kubernetes. In order to make a reliable extension mechanism for admission,
+we must reconsider the current chain to produce understandable extensions.
+
+We are proposing a series of changes to admission control in an attempt to make
+it less error-prone, easier to configure, and more flexible for admission plugin
+developers.
+
+## Requirements for extensible admission
+Ordering the admission chain is critical.  One mutating admission plugin can invalidate
+the results of other admission plugins.  For instance, the `LimitVerifier` would
+invalidate results for the `ResourceQuota` plugin.  More generally, object mutations
+must happen before object validations.  In addition, compiled validating admission plugins
+must run after mutating admission webhooks.  Splitting the chain into mutators and validators
+will resolve the coarse grained ordering problem and would require two different webhook APIs
```

Comment:
> would require two different webhook APIs

The differences don't seem significant enough to warrant two separate resources. I'd expect mutating webhooks to indicate that in their spec, and the portions of the status related to mutation simply be required to be empty in responses from non-mutating webhooks.

Reply:
> The differences don't seem significant enough to warrant two separate resources. I'd expect mutating webhooks to indicate that in their spec, and the portions of the status related to mutation simply be required to be empty in responses from non-mutating webhooks.

I really like different things to be different.  You just described different validation requirements.  They are definitely similar, but the semantics are definitely different.

Reply:
Clarify: Are you loosely describing initializers as "webhooks", or are you introducing a new extension mechanism, which is a webhook that can also mutate?

Reply:
> Clarify: Are you loosely describing initializers as "webhooks", or are you introducing a new extension mechanism, which is a webhook that can also mutate?

Webhooks already exist.  They are one of the admission features that we're trying to move to beta.  Right now they aren't mutating, but to have enough expressablity to support our current chain, they need to add it.